core/arch/riscv/mm/core_mmu_arch.c

```diff
@@ -0,0 +1,714 @@
+				   struct tee_mmap_region *mm)
+{
+	size_t n = 0;
+	void *pgt = core_mmu_get_root_pgt_va(prtn);
+
+	memset(pgt, 0, RISCV_MMU_PGT_SIZE);
+	memset(prtn->pool_pgts, 0, RISCV_MMU_MAX_PGTS * RISCV_MMU_PGT_SIZE);
+
+	for (n = 0; !core_mmap_is_end_of_table(mm + n); n++)
+		if (!core_mmu_is_dynamic_vaspace(mm + n))
+			core_mmu_map_region(prtn, mm + n);
+}
+
+void tlbi_mva_range(vaddr_t va, size_t len,
+		    size_t granule)
+{
+	assert(granule == CORE_MMU_PGDIR_SIZE || granule == SMALL_PAGE_SIZE);
+	assert(!(va & (granule - 1)) && !(len & (granule - 1)));
+
+	while (len) {
+		tlbi_mva_allasid_nosync(va);
+		len -= granule;
+		va += granule;
+	}
+}
+
+void tlbi_mva_range_asid(vaddr_t va, size_t len,
+			 size_t granule, uint32_t asid)
+{
+	assert(granule == CORE_MMU_PGDIR_SIZE || granule == SMALL_PAGE_SIZE);
+	assert(!(va & (granule - 1)) && !(len & (granule - 1)));
+
+	while (len) {
+		tlbi_mva_asid_nosync(va, asid);
+		len -= granule;
+		va += granule;
+	}
+}
+
+TEE_Result cache_op_inner(enum cache_op op, void *va, size_t len)
+{
+	switch (op) {
+	case DCACHE_CLEAN:
+		dcache_op_all(DCACHE_OP_CLEAN);
+		break;
+	case DCACHE_AREA_CLEAN:
+		dcache_clean_range(va, len);
+		break;
+	case DCACHE_INVALIDATE:
+		dcache_op_all(DCACHE_OP_INV);
+		break;
+	case DCACHE_AREA_INVALIDATE:
+		dcache_inv_range(va, len);
+		break;
+	case ICACHE_INVALIDATE:
+		icache_inv_all();
+		break;
+	case ICACHE_AREA_INVALIDATE:
+		icache_inv_range(va, len);
+		break;
+	case DCACHE_CLEAN_INV:
+		dcache_op_all(DCACHE_OP_CLEAN_INV);
+		break;
+	case DCACHE_AREA_CLEAN_INV:
+		dcache_cleaninv_range(va, len);
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+	return TEE_SUCCESS;
+}
+
+unsigned int asid_alloc(void)
+{
+	uint32_t exceptions = cpu_spin_lock_xsave(&g_asid_spinlock);
+	unsigned int r = 0;
+	int i = 0;
+
+	bit_ffc(&g_asid, BIT(g_asid), &i);
+	if (i == -1) {
+		r = 0;
+	} else {
+		bit_set(&g_asid, i);
+		r = i + 1;
+	}
+
+	cpu_spin_unlock_xrestore(&g_asid_spinlock, exceptions);
+
+	return r;
+}
+
+void asid_free(unsigned int asid)
+{
+	uint32_t exceptions = cpu_spin_lock_xsave(&g_asid_spinlock);
+
+	/* Only even ASIDs are supposed to be allocated */
+	assert(!(asid & 1));
+
+	if (asid) {
+		int i = asid - 1;
+
+		assert(i < BIT(g_asid) && bit_test(&g_asid, i));
+		bit_clear(&g_asid, i);
+	}
+
+	cpu_spin_unlock_xrestore(&g_asid_spinlock, exceptions);
+}
+
+bool arch_va2pa_helper(void *va, paddr_t *pa)
+{
+	vaddr_t vaddr = (vaddr_t)va;
+	struct mmu_pgt *pgt = NULL;
+	struct mmu_pte *pte = NULL;
+	int level = 0;
+	unsigned int idx = 0;
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+
+	if (!pa)
+		return false;
+
+	pgt = core_mmu_get_root_pgt_va(prtn);
+
+	for (level = CORE_MMU_BASE_TABLE_LEVEL; level >= 0; level--) {
+		idx = core_mmu_pgt_idx(vaddr, level);
+		pte = core_mmu_table_get_entry(pgt, idx);
+
+		if (core_mmu_entry_is_invalid(pte)) {
+			return false;
+		} else if (core_mmu_entry_is_leaf(pte)) {
+			*pa = pte_to_pa(pte);
```

Comment:
Should add page offset ?

```suggestion
			*pa = pte_to_pa(pte) |
				(vaddr & (BIT64(RISCV_PGSHIFT) - 1));
```